src/Workspaces/Remote/Core/Services/AssetService.cs

```diff
@@ -172,6 +239,88 @@ public Entry(object @object)
                 Object = @object;
             }
         }
+
+        private struct ChecksumSynchronizer
+        {
+            private readonly AssetService _assetService;
+
+            public ChecksumSynchronizer(AssetService assetService)
+            {
+                _assetService = assetService;
+            }
+
+            public async Task SynchronizeAsync(Checksum solutionChecksum, CancellationToken cancellationToken)
+            {
+                // get children of solution checksum object at once
+                var solutionChecksums = new HashSet<Checksum>();
+                var solutionChecksumObject = await _assetService.GetAssetAsync<SolutionChecksumObject>(solutionChecksum, cancellationToken).ConfigureAwait(false);
+
+                AppendIfNeeded(solutionChecksums, solutionChecksumObject.Info);
+                AppendIfNeeded(solutionChecksums, solutionChecksumObject.Projects);
+
+                await _assetService.SynchronizeAssetsAsync(solutionChecksums, cancellationToken).ConfigureAwait(false);
+
+                // get children of project checksum objects at once
+                var projectChecksums = new HashSet<Checksum>();
+                foreach (var projectChecksum in solutionChecksumObject.Projects)
```

Comment:
it looks like you're looping over projects twice (hre and below).  Why not merge the loops?